### DISCUSSION TITLE
how about adding reload feature for axum?

### DISCUSSION THREAD

**Discussion by zishon** [tokio-rs/axum#2161] [Q&A] [answered]
I have read all the crate document and post here
I don't find feature such as nginx have: nginx -s reload
I think this feature is useful, how about adding reload feature for axum in future version?

**Comment by davidpdrsn** [comment]
What does `nginx -s reload` do?

**Reply by zishon** [reply]
It starts new workers, and old workers continue process connections which already accepted.
It does not stop the service while loading new configs.

**Comment by jplatte** [comment]
Hot reloading when the code changes can be achieved using [`listenfd`](https://crates.io/crates/listenfd) + systemfd + cargo-watch.

Maybe we should add an example for that. @davidpdrsn wdyt?

**Reply by frenetisch-applaudierend** [reply]
I think the requested feature is different from hot reload, as it relates to a production setting. From the [NGINX FAQ](https://www.nginx.com/faq/how-does-zero-downtime-configuration-testingreload-in-nginx-plus-work/):

> You can reload the NGINX configuration without any downtime, and without interrupting any current connections. A soft reconfiguration spawns a new set of NGINX worker processes, and the old ones continue to work until completion (but do not accept new requests – all new requests are accepted by the new set of worker processes).

Not sure how this would translate to axum though?

**Reply by jplatte** [reply]
Well, axum doesn't have a standardized configuration scheme, so it can't have functionality to reload configuration. That's why I guessed that @zishon wanted hot reloading on code change. Let's wait for clarification 🤷🏼

**Reply by zishon** [reply]
what @frenetisch-applaudierend say is exactly what I expect.
scenario here is:
when I update the config of an axum project on prod server, I want to reload the service without introduce any downtime.

Suppose I have an application called 'axum-demo' and a configuration file named 'config.toml.' This configuration file contains a cache duration setting 'cache_ttl = 600.' Now, I need to change it to 'cache_ttl = 30.' After making this modification, I want the 'axum-demo' application to reflect these changes without using the 'shutdown + start' method, as this would cause a service interruption (because it will be down for a few seconds).

**Reply by jplatte** [reply]
Can you show the code you use to parse the configuration file at startup, and how do you pass that to handlers? Are you using [`State`](https://docs.rs/axum/latest/axum/#using-the-state-extractor)? If yes, what does your state type look like?

**Reply by zishon** [reply]
I don't think this is code relevant mechanisms.
Set up a standardized configuration scheme for axum may be a solution.

**Reply by jplatte** [reply]
No, we are not going to add configuration to axum. The way you read your config file now isn't required for me to make an example of how to reload, but it will make it easier and I'm helping you voluntarily afterall.

**Reply by zishon** [reply]
```rust
use std::fmt::format;

use axum::{Extension, Router};
use axum::extract::State;
use axum::response::{IntoResponse, Response};
use axum::routing::{get, post};
use toml::Value;

#[derive(Debug, Clone)]
struct AppConfig {
    app_id: String,
    app_secret: String,
}

#[derive(Debug, Clone)]
struct User {
    name: String,
}

impl IntoResponse for User {
    fn into_response(self) -> Response {
        format!("User Name: {}", self.name).into_response()
    }
}

async fn hello(State(app_config): State<AppConfig>) -> User {
    //get user information from user service by using app_id and sign
    println!("{}", app_config.app_id);
    println!("{}", app_config.app_secret);
    User {
        name: "Demo".to_owned(),
    }
}

#[tokio::main]
async fn main() {
    let raw_config = tokio::fs::read_to_string("config.toml")
        .await
        .expect("Failed to read config.toml");
    let parsed_config: Value = toml::from_str(&raw_config)
        .expect("Failed to parse config.toml");

    let app_id = parsed_config.get("app_id").expect("Failed to read App ID");
    let app_secret = parsed_config.get("app_secret").expect("Failed to read App Secret");

    let app_config = AppConfig {
        app_id: app_id.to_string(),
        app_secret: app_secret.to_string(),
    };

    let app = Router::new()
        .route("/hello", get(hello))
        .with_state(app_config);

    // run it with hyper on localhost:3000
    axum::Server::bind(&"0.0.0.0:3000".parse().unwrap())
        .serve(app.into_make_service())
        .await
        .unwrap();
}
```

**Reply by jplatte** [reply, marked as answer]
Alright, so it's just a small struct that is cloned everywhere and parsed using the `toml` crate. In that case, I would recommend making your state type

```rust
use std::sync::RwLock;

struct AppState {
    // Could also use the arc_swap crate for Arc<ArcSwap<Arc<AppConfig>>>
    config: Arc<RwLock<AppConfigInner>>,
}

// With the derived Deserialize impl, you can use
// toml::from_str::<AppConfigInner>(&raw_config)
// directly, instead of going through toml::Value
#[derive(Deserialize)]
struct AppConfig {
    app_id: String,
    app_secret: String,
}

// To allow extracting State<AppConfig> but supplying `.with_state(<AppState instance>)`
impl FromRef<AppState> for AppConfig {
    fn from_ref(state: &AppState) -> Self {
        state.config.read().unwrap().clone()
    }
}
```

then you can share a `clone` of the `Arc<RwLock<AppConfigInner>>` with whatever update mechanism you want for the config.

For unix-like systems, I would recommend you hook up the `SIGHUP` signal, which is a (semi-?)standard way of signalling a process that it should reload its configuration. You could spawn a background tokio task with `tokio::spawn` that checks for `Signal::Hup` using the [async-signal](https://docs.rs/async-signal/0.2.0/async_signal/) crate, and once it has received that signal, it reloads the config file and writes it to the shared `Arc<RwLock<AppConfig>>` using `*shared_config.write().unwrap() = new_config`.
To send this signal, you can use `pkill -SIGHUP <name of the executable>` or `kill -SIGHUP <process ID>`.

Feel free to ask any follow-up questions if the above isn't totally clear.

**Reply by zishon** [reply]
So appreciated for your help.

**Comment by zishon** [comment]


**Reply by jplatte** [reply]
Could you please answer to the thread that has the message you're quoting, instead of posting a new answer?

About your main question: as stated earlier, axum has no standardized configuration scheme so how could it possibly have configuration reloading? If you can post the existing code you have for config loading though, maybe we can still help.

**Comment by davidpdrsn** [comment]
> Maybe we should add an example for that. @davidpdrsn wdyt?

Yeah that sounds like a good idea! I've never set something like that up myself.

**Comment by dennj** [comment]
This enables Axum hot reload:
[axum-hotreload](https://github.com/dennj/axum-hotreload)
